src/refresh-runtime.js

```diff
@@ -581,41 +581,60 @@ function debounce(fn, delay) {
   };
 }
 
-const enqueueUpdate = debounce(performReactRefresh, 16);
+const enqueueUpdate = debounce(() => {
+  window.__beforePerformReactRefresh?.();
+  performReactRefresh();
+}, 16);
 
 export function validateRefreshBoundaryAndEnqueueUpdate(
+  id,
   prevExports,
   nextExports,
 ) {
-  if (!predicateOnExport(prevExports, (key) => key in nextExports)) {
+  const ignoredExports = window.__getReactRefreshIgnoredExports?.({ id }) ?? [];
+  if (
+    predicateOnExport(
+      ignoredExports,
+      prevExports,
+      (key) => key in nextExports,
+    ) !== true
+  ) {
     return "Could not Fast Refresh (export removed)";
   }
-  if (!predicateOnExport(nextExports, (key) => key in prevExports)) {
+  if (
+    predicateOnExport(
+      ignoredExports,
+      nextExports,
+      (key) => key in prevExports,
+    ) !== true
+  ) {
     return "Could not Fast Refresh (new export)";
   }
 
   let hasExports = false;
   const allExportsAreComponentsOrUnchanged = predicateOnExport(
+    ignoredExports,
     nextExports,
     (key, value) => {
       hasExports = true;
       if (isLikelyComponentType(value)) return true;
       return prevExports[key] === nextExports[key];
     },
   );
-  if (hasExports && allExportsAreComponentsOrUnchanged) {
+  if (hasExports && allExportsAreComponentsOrUnchanged === true) {
     enqueueUpdate();
   } else {
-    return "Could not Fast Refresh. Learn more at https://github.com/vitejs/vite-plugin-react-swc#consistent-components-exports";
+    return `Could not Fast Refresh (${allExportsAreComponentsOrUnchanged} export is incompatible). Learn more at https://github.com/vitejs/vite-plugin-react-swc#consistent-components-exports`;
```

Comment:
```suggestion
    return `Could not Fast Refresh ("${allExportsAreComponentsOrUnchanged}" export is incompatible). Learn more at https://github.com/vitejs/vite-plugin-react-swc#consistent-components-exports`;
```
I think this would reduce ambiguity.